### PR TITLE
[datasets] Use generators for merge stage in push-based shuffle

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

This uses the generators introduced in #25247 to reduce memory usage during the merge stage in push-based shuffle. These tasks merge groups of map outputs, so it fits a generator pattern where we want to return merged outputs one at a time. Verified that this allows for merging more/larger objects at a time than the current list-based version.

I also tried this for the map stage in random_shuffle, but it didn't seem to make a difference in memory usage for Arrow blocks. I think this is probably because Arrow is already doing some zero-copy optimizations when selecting rows?

Also adds a new line to Dataset stats for memory usage. Unfortunately it's hard to get an accurate reading of physical memory usage in Python and this value will probably be an overestimate in a lot of cases. I didn't see a difference before and after this PR for the merge stage, for example. Arguably this field should be opt-in. For 100MB partitions, for example:
```
        Substage 0 read->random_shuffle_map: 10/10 blocks executed
        * Remote wall time: 1.44s min, 3.32s max, 2.57s mean, 25.74s total
        * Remote cpu time: 1.42s min, 2.53s max, 2.03s mean, 20.25s total
        * Worker memory usage (MB): 462 min, 864 max, 552 mean
        * Output num rows: 12500000 min, 12500000 max, 12500000 mean, 125000000 total
        * Output size bytes: 101562500 min, 101562500 max, 101562500 mean, 1015625000 total
        * Tasks per node: 10 min, 10 max, 10 mean; 1 nodes used

        Substage 1 random_shuffle_reduce: 10/10 blocks executed
        * Remote wall time: 1.47s min, 2.94s max, 2.17s mean, 21.69s total
        * Remote cpu time: 1.45s min, 1.88s max, 1.71s mean, 17.09s total
        * Worker memory usage (MB): 462 min, 1047 max, 831 mean
        * Output num rows: 12500000 min, 12500000 max, 12500000 mean, 125000000 total
        * Output size bytes: 101562500 min, 101562500 max, 101562500 mean, 1015625000 total
        * Tasks per node: 10 min, 10 max, 10 mean; 1 nodes used
```

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/impl/stats.py [hunk 1]
```diff
@@ -315,6 +315,14 @@ def _summarize_blocks(self, blocks: List[BlockMetadata], is_substage: bool) -> s
                 fmt(sum([e.cpu_time_s for e in exec_stats])),
             )
 
+            out += indent
+            memory_stats = [int(e.max_rss_bytes / 1e6) for e in exec_stats]
```
Comment (ericl) [inline python/ray/data/impl/stats.py:319]:
```suggestion
            memory_stats = [round(e.max_rss_bytes / 1024 * 1024, 2) for e in exec_stats]
```

File: python/ray/data/impl/stats.py [hunk 2]
```diff
@@ -315,6 +315,14 @@ def _summarize_blocks(self, blocks: List[BlockMetadata], is_substage: bool) -> s
...
+            out += "* Worker memory usage (MB): {} min, {} max, {} mean\n".format(
```
Comment (ericl) [inline python/ray/data/impl/stats.py:320]:
```suggestion
            out += "* Worker memory usage (MiB): {} min, {} max, {} mean\n".format(
```

**Comment by stephanie-wang** [comment]
Need to fix the CI error after the stats formatting change, but this is ready for another look.

File: python/ray/data/_internal/stats.py [hunk 3]
```diff
@@ -315,6 +316,14 @@ def _summarize_blocks(self, blocks: List[BlockMetadata], is_substage: bool) -> s
                 fmt(sum([e.cpu_time_s for e in exec_stats])),
             )
 
+            out += indent
+            memory_stats = [round(e.max_rss_bytes / 1024 * 1024, 2) for e in exec_stats]
+            out += "* Worker memory usage (MiB): {} min, {} max, {} mean\n".format(
```
Comment (ericl) [inline python/ray/data/_internal/stats.py:321]:
```suggestion
            out += "* Peak heap memory usage (MiB): {} min, {} max, {} mean\n".format(
```

File: python/ray/data/block.py [hunk 4]
```diff
@@ -114,6 +119,9 @@ def __init__(self):
         self.wall_time_s: Optional[float] = None
         self.cpu_time_s: Optional[float] = None
         self.node_id = ray.runtime_context.get_runtime_context().node_id.hex()
+        # Max memory usage. May be an overestimate since we do not
+        # differentiate from previous tasks on the same worker.
+        self.max_rss_bytes: int = 0
```
Comment (ericl) [inline python/ray/data/block.py:124]: Hmm could the peak usage actually be higher during the middle of the task?

**Review by ericl — APPROVED** [review]

File: python/ray/data/block.py (hunk 4 above)
Comment (ericl) [reply python/ray/data/block.py:122]: Maybe we should also record this at the start, so we know if the task increased it.

File: python/ray/data/block.py [hunk 5]
```diff
@@ -1,32 +1,35 @@
-from dataclasses import dataclass
+import resource
```
Comment (matthewdeng) [inline python/ray/data/block.py:1]: @stephanie-wang looks like this is causing Windows tests to fail!

Comment (stephanie-wang) [reply python/ray/data/block.py:1]: Sorry about that :( I forgot to remove the tests-ok label!

**Pull request merged**